File: solutions/_93_restore_ip_addresses/solution.py
from typing import List


class Solution:
    def restoreIpAddresses(self, s: str) -> List[str]:
        def backtrack(start, path):
            if len(path) == 4:
                if start == len(s):
                    res.append(".".join(path))
                return
            for i in range(1, 4):
                if start + i > len(s):
                    break
                if i > 1 and s[start] == "0":
                    break
                octet = int(s[start : start + i])
                if octet < 256:
                    backtrack(start + i, path + [s[start : start + i]])

        res = []
        backtrack(0, [])
        return res


# https://leetcode.com/problems/restore-ip-addresses/

File: solutions/_93_restore_ip_addresses/solution.java

package _93_restore_ip_addresses;

import java.util.ArrayList;
import java.util.List;

class Solution {
    String s; // input string
    List<String> res = new ArrayList<>(); // result
    List<String> path = new ArrayList<>(); // current path

    public List<String> restoreIpAddresses(String s) {
        this.s = s;
        backtrack(0);
        return res;
    }

    private void backtrack(int start) {
        if (path.size() == 4) {
            if (start == s.length()) {
                res.add(String.join(".", path));
            }
        } else {
            for (int i = 1; i <= 3; i++) {
                if (start + i > s.length())
                    break;
                if (i > 1 && s.charAt(start) == '0')
                    break;
                String octet = s.substring(start, start + i);
                if (Integer.parseInt(octet) <= 255) {
                    path.add(octet);
                    backtrack(start + i);
                    path.remove(path.size() - 1);
                }
            }
        }
    }
}

/*
 * https://leetcode.com/problems/restore-ip-addresses/
 */